### PR TITLE
fix: replaced OCL from web to JSON (in memory)

### PR DESCRIPTION
**Pull request by ajinkyagadewar**
## Type of PR (check all applicable)

- [ ] 💼 New Package
- [x] ♻️ Refactor
- [ ] ✨ New Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🧪 Tests
- [ ] 🔖 Release
- [ ] 🚩 Other

## Description
This PR replaces the implementation of Terminology Services from OCL web to locally downloaded OCL JSONs

## Tickets & Related Documents
(https://app.clickup.com/t/86bzrgeut)

## Added code snippets?
- [ ] 👍 Yes
- [x] 🙅 No, because they aren't needed

## Added tests?

- [ ] 👍 Yes
- [ ] 🙅 No, because they aren't needed
- [ ] 🙋 No, because I need help

## Added to documentation?

- [ ] 📜 readme
- [ ] 📓 general documentation
- [x] 🙅 no documentation needed

### No docs? Add a note
We have downloaded JSON files from OCL website and stored it in the repo. This doesn't require any documentation.

## Are there any post-deployment tasks we need to perform?
Yes. Please destroy, build and initiate the shr-mediator.
-
-
-

### CODE REVIEW

**Review by marrouchi — CHANGES_REQUESTED** [review]
Let's add configs to the package folder under "config" and add them to docker-compose.yml.

**Comment by ajinkyagadewar** [comment]
> Let's add configs to the package folder under "config" and add them to docker-compose.yml.

Added

File: projects/shr-mediator/src/botswana/services/terminology.service.ts [hunk 1]
```diff
@@ -2,15 +2,38 @@ import { R4 } from '@ahryman40k/ts-fhir-types';
...
+  private pimsToIpmsMappings: any[];
+  private ipmsCodingInfo: any;
+
+  constructor(private readonly logger: LoggerService) {
+    this.loadJsonData();
+  }
+
+  private loadJsonData() {
+    try {
+      const dataPath = path.resolve(__dirname, '..', 'ocl_json');
```
Comment (marrouchi) [inline projects/shr-mediator/src/botswana/services/terminology.service.ts:23]: Can we do some refactoring ?
```
const loadJsonFile = async (fileName) => {
    try {
      const dataPath = path.resolve(__dirname, '..', 'ocl_json');
      const filePath = path.join(dataPath, fileName);
      const fileContents = await fs.promises.readFile(filePath, 'utf8');
      return JSON.parse(fileContents);
    } catch (error) {
      this.logger.error(`Error loading JSON data from ${fileName}: `, error);
      return null;  // or return an empty array/object based on expected data structure
    }
  };

  this.ipmsToPimsMappings = await loadJsonFile('IPMSLAB_to_PIMSLAB.json') || {};
  this.ipmsToCielMappings = await loadJsonFile('IPMSLAB_to_CIEL.json') || {};
  this.pimsToIpmsMappings = await loadJsonFile('PIMSLAB_to_IPMSLAB.json') || {};
  this.ipmsCodingInfo = await loadJsonFile('IPMS_Coding.json') || [];

  if (!Array.isArray(this.ipmsCodingInfo)) {
    this.logger.error('IPMS Coding info is not an array');
    this.ipmsCodingInfo = [];
  }
```

File: projects/shr-mediator/src/botswana/services/terminology.service.ts [hunk 2]
```diff
@@ -148,44 +161,35 @@ export class TerminologyService {
     }
   }
 
-  async getIpmsCode(q: string, c = '') {
+  getIpmsCodeFromMemory(mappings: any[], code: string) {
```
Comment (marrouchi) [inline projects/shr-mediator/src/botswana/services/terminology.service.ts:164]: Can we enforce typing and avoid "any" ?

This allows to generate automatically typscript from json : https://transform.tools/json-to-typescript

File: projects/shr-mediator/src/botswana/services/terminology.service.ts [hunk 3]
```diff
@@ -148,44 +161,35 @@ export class TerminologyService {
...
-        // this.logger.log(`IPMS Coding Info: ${JSON.stringify(ipmsCodingInfo)}`)
+        const ipmsCode = mappings[mappingIndex].from_concept_code;
+        const ipmsDisplay = mappings[mappingIndex].from_concept_name_resolved;
+        const ipmsCodingInfoID = this.ipmsCodingInfo.find((concept: any) => concept.id === ipmsCode);
+
         let ipmsMnemonic, hl7Flag;
-        if (ipmsCodingInfo) {
-          ipmsMnemonic = ipmsCodingInfo.names.find(
+        if (ipmsCodingInfoID) {
+          ipmsMnemonic = ipmsCodingInfoID.names.find(
```
Comment (marrouchi) [inline projects/shr-mediator/src/botswana/services/terminology.service.ts:184]: Potential for "undefined" error

File: projects/shr-mediator/src/botswana/services/terminology.service.ts [hunk 4]
```diff
@@ -201,51 +205,25 @@ export class TerminologyService {
...
-      if (codeMapping && codeMapping.length > 0) {
+      if (mapping) {
         return {
-          code: codeMapping[0].to_concept_code,
-          display: codeMapping[0].to_concept_name_resolved,
+          code: mapping.to_concept_code,
+          display: mapping.to_concept_name_resolved,
         };
       } else {
         return {};
```
Comment (marrouchi) [inline projects/shr-mediator/src/botswana/services/terminology.service.ts:219]: If mapping is not found, let's throw and exception or return undefined or null

File: projects/shr-mediator/src/botswana/services/terminology.service.ts [hunk 5]
```diff
@@ -201,51 +205,25 @@ export class TerminologyService {
     }
   }
 
-  async getMappedCode(q: string): Promise<any> {
+  getMappedCodeFromMemory(mappings: any[], code: string): any {
     try {
-      const codeMapping = await this.getOclMapping(q);
-
-      //this.logger.log(`Code Mapping: ${JSON.stringify(codeMapping)}`)
+      this.logger.log(`Code: ${code}`)
```
Comment (marrouchi) [inline projects/shr-mediator/src/botswana/services/terminology.service.ts:210]:
```suggestion
      this.logger.log(`Terminology mapping for code: ${code}`)
```

**Review by marrouchi — CHANGES_REQUESTED** [review]

File: projects/shr-mediator/src/botswana/services/terminology.service.ts (hunk 4 above)
Comment (ajinkyagadewar) [reply projects/shr-mediator/src/botswana/services/terminology.service.ts:230]: Fixed

File: projects/shr-mediator/src/botswana/services/terminology.service.ts (hunk 3 above)
Comment (ajinkyagadewar) [reply projects/shr-mediator/src/botswana/services/terminology.service.ts:195]: Fixed

File: projects/shr-mediator/src/botswana/services/terminology.service.ts (hunk 2 above)
Comment (ajinkyagadewar) [reply projects/shr-mediator/src/botswana/services/terminology.service.ts]: Fixed

**Pull request merged**